src/flamegpu/simulation/CUDAEnsemble.cu

```diff
@@ -150,14 +151,26 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector& plans) {
         }
     }
 
+    // Select the actual devices to be used, based on user provided gpus, architecture compatibility, and optionally mpi ranks per node.
+    // For non-mpi builds / configurations, just use all the devices provided by the user / all visible devices (then check they are valid later)
+    // For MPI builds with mpi enabled, load balance the gpus across mpi ranks within the shared memory system. If there are more ranks than gpus, latter ranks will not participate.
     std::set<int> devices;
+    // initialise the local devices set to be the non-mpi behaviour, using config.devices or all visible cuda devices
     if (config.devices.size()) {
         devices = config.devices;
     } else {
+        // If no devices were specified by the user, use all visible devices but load balance if MPI is in use.
         for (int i = 0; i < device_count; ++i) {
             devices.emplace(i);
         }
     }
+#ifdef FLAMEGPU_ENABLE_MPI
+    // if MPI is enabled at compile time, and mpi is enabled in the config, use the MPIEnsemble method to assign devices balanced across ranks
```

Comment:
```suggestion
    // if MPI is enabled at compile time, use the MPIEnsemble method to assign devices balanced across ranks
```

config mpi is compile time const

Reply:
Based on this. 

https://github.com/FLAMEGPU/FLAMEGPU2/blob/cdad83a2dd9c34e2ebefc3a7fbea740180e7a62e/src/flamegpu/simulation/CUDAEnsemble.cu#L66